src/Raven.Server/Documents/Sharding/BucketStatsHolder.cs

```diff
@@ -0,0 +1,119 @@
+            _mergedChangeVectors = new Dictionary<int, ChangeVector>();
+        }
+
+        public void UpdateBucket(int bucket, long nowTicks, long sizeChange, long numOfDocsChanged)
+        {
+            _values.TryGetValue(bucket, out var bucketStats);
+
+            bucketStats.Size += sizeChange;
+            bucketStats.NumberOfDocuments += numOfDocsChanged;
+            bucketStats.LastModifiedTicks = nowTicks;
+
+            _values[bucket] = bucketStats;
+        }
+
+        public void UpdateBucket(int bucket, long nowTicks, long sizeChange, long numOfDocsChanged, int changeVectorIndex, ref TableValueReader value)
+        {
+            UpdateBucket(bucket, nowTicks, sizeChange, numOfDocsChanged);
+
+            var changeVector = TableValueToChangeVector(changeVectorIndex, ref value);
+            if (_mergedChangeVectors.TryGetValue(bucket, out var currentMergedCv))
+            {
+                changeVector = currentMergedCv.MergeWith(changeVector, _ctx);
+            }
+
+            _mergedChangeVectors[bucket] = changeVector;
+        }
+
+        public void UpdateBucketStatsTreeBeforeCommit(Transaction tx)
+        {
+            int bucketStatsSize = sizeof(Documents.BucketStats);
+            var tree = tx.ReadTree(ShardedDocumentsStorage.BucketStatsSlice);
+
+            foreach ((int bucket, Documents.BucketStats inMemoryStats) in _values)
+            {
+                using (tx.Allocator.Allocate(sizeof(int), out var keyBuffer))
+                {
+                    *(int*)keyBuffer.Ptr = bucket;
+                    var keySlice = new Slice(keyBuffer);
+                    var readResult = tree.Read(keySlice);
+
+                    Documents.BucketStats stats;
+                    IDisposable cvScope = null;
+                    Slice cvSlice = default;
+
+                    if (readResult == null)
+                    {
+                        stats = inMemoryStats;
+                    }
+                    else
+                    {
+                        stats = *(Documents.BucketStats*)readResult.Reader.Base;
+                        stats.Size += inMemoryStats.Size;
+                        stats.NumberOfDocuments += inMemoryStats.NumberOfDocuments;
+                        stats.LastModifiedTicks = inMemoryStats.LastModifiedTicks;
+
+                        cvScope = stats.GetMergedChangeVector(tx.Allocator, readResult.Reader, out cvSlice);
+                    }
+
+                    if (stats.Size == 0 && stats.NumberOfDocuments == 0)
+                    {
+                        tree.Delete(keySlice);
+                        continue;
+                    }
+
+                    if (_mergedChangeVectors.TryGetValue(bucket, out var changeVector))
+                    {
+                        using (cvScope)
+                        {
+                            changeVector = changeVector.MergeWith(cvSlice.HasValue ? cvSlice.ToString() : null, _ctx);
```

Comment:
let's create an overload for `stats.GetMergedChangeVector` to return it as string, so we can avoid some of the scopes here.

Reply:
done